### PR TITLE
Aoreshko/Madport-12-0001

### PR DESCRIPTION
**Pull request by AlexanderOreshko**
Patching o3de engine to build with o3de-gruber.
Methods wrapped to CARBONATED macro are added to:
1. AzFramework/Components/TransformComponent : SetClientSimulated
2. PhysXCharacters/Components/CharacterControllerComponent: SetMaterialByName & SetTag
3. AzCore/Math/Vector2 : Rotate
AzCore/Math/Quaternion: Multiply (*) Vector3 operator

### CODE REVIEW

File: Code/Legacy/CryCommon/Cry_Color.h [hunk 1]
```diff
@@ -433,3 +433,112 @@ inline void Color_tpl<T>::clamp(T bottom, T top)
 
 #define Col_TrackviewDefault    ColorF (0.187820792f, 0.187820792f, 1.0f)
 #define Clr_Empty                               ColorF(0.0f, 0.0f, 0.0f, 1.0f)
+
```
Comment (alexeykostin) [inline Code/Legacy/CryCommon/Cry_Color.h]: Please move these changes to the game code in `o3de-gruber\Code\Legacy\CryCommonAddon\CryCommon`. You may want to create a new include header, like `Cry_ColorCompat.h`.

File: Code/Framework/AzCore/AzCore/Component/TransformBus.h [hunk 2]
```diff
@@ -302,6 +302,13 @@ namespace AZ
 
         //! Set the behavior at runtime when this entity's parent's transform changes.
         virtual void SetOnParentChangedBehavior([[maybe_unused]] OnParentChangedBehavior onParentChangedBehavior) {}
+
+        // Gruer patch begin aoreshko: MADPORT-12
```
Comment (alexeykostin) [inline Code/Framework/AzCore/AzCore/Component/TransformBus.h]: Please use the proper markup for O3DE patches

**Review by alexeykostin — CHANGES_REQUESTED** [review]

File: Code/Framework/AzCore/AzCore/Math/Quaternion.h [hunk 3]
```diff
@@ -228,6 +228,13 @@ namespace AZ
         bool operator==(const Quaternion& rhs) const;
         bool operator!=(const Quaternion& rhs) const;
 
+        // Gruer patch begin aoreshko: MADPORT-12
```
Comment (vmednonogov) [inline Code/Framework/AzCore/AzCore/Math/Quaternion.h]: // Gru(b)er -> // carbonated 
here and in other places

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.h [hunk 4]
```diff
@@ -125,6 +125,13 @@ namespace PhysX
         float GetHalfForwardExtent() override;
         void SetHalfForwardExtent(float halfForwardExtent) override;
 
+        // carbonated begin aoreshko game_specific_1
+#if defined(CARBONATED)
+        void SetMaterialByName(uint32_t index, const AZStd::string& name) override;
+        void SetTag(const AZ::Crc32& tag) override;
+#endif
+        // carbonated end game_specific_1
```
Comment (vmednonogov) [inline Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.h]: // carbonated end -> // carbonated end aoreshko
here and in other places

File: Code/Framework/AzCore/AzCore/Math/Quaternion.h (hunk 3 above)
Comment (vmednonogov) [inline Code/Framework/AzCore/AzCore/Math/Quaternion.h:231]: // Gruer -> // Gruber
there and in other places

File: Code/Framework/AzCore/AzCore/Math/Quaternion.cpp [hunk 5]
```diff
@@ -517,4 +517,14 @@ namespace AZ
             return halfAngle * 2.0f * (imaginary / length);
         }
     }
+
+    // carbonated begin aoreshko game_specific_1
```
Comment (aefimov3) [inline Code/Framework/AzCore/AzCore/Math/Quaternion.cpp]: Please add #if defined(CARBONATED)

File: Code/Framework/AzCore/AzCore/Math/Quaternion.h [hunk 6]
```diff
@@ -228,6 +228,13 @@ namespace AZ
         bool operator==(const Quaternion& rhs) const;
         bool operator!=(const Quaternion& rhs) const;
 
+        // carbonated begin aoreshko game_specific_1
```
Comment (aefimov3) [inline Code/Framework/AzCore/AzCore/Math/Quaternion.h]: Please add #if defined(CARBONATED)

File: Code/Framework/AzCore/AzCore/Math/Vector2.h [hunk 7]
```diff
@@ -163,6 +163,18 @@ namespace AZ
         //! The positive rotation direction is defined such that the x-axis is rotated into the y-axis.
         Vector2 GetPerpendicular() const;
 
+        // carbonated begin aoreshko game_specific_1
```
Comment (aefimov3) [inline Code/Framework/AzCore/AzCore/Math/Vector2.h]: According to the document we skip names here, just the feature name. We keep names in a document registry only.
Please remove aoreshko,

File: Code/Framework/AzFramework/AzFramework/Components/TransformComponent.cpp [hunk 8]
```diff
@@ -707,6 +707,16 @@ namespace AzFramework
         provided.push_back(AZ_CRC("TransformService", 0x8ee22c50));
     }
 
+     // carbonated begin aoreshko game_specific_1
```
Comment (aefimov3) [inline Code/Framework/AzFramework/AzFramework/Components/TransformComponent.cpp]: Please remove aoreshko.

File: Code/Framework/AzFramework/AzFramework/Components/TransformComponent.h [hunk 9]
```diff
@@ -74,6 +74,13 @@ namespace AzFramework
         //! This will use worldTM as a localTM and move the transform relative to the parent.
         void SetParentRelative(AZ::EntityId id) override;
 
+         // carbonated begin aoreshko game_specific_1
```
Comment (aefimov3) [inline Code/Framework/AzFramework/AzFramework/Components/TransformComponent.h]: Please remove aoreshko.

File: Code/Framework/AzFramework/AzFramework/Components/TransformComponent.h [hunk 10]
```diff
@@ -190,5 +197,11 @@ namespace AzFramework
         bool m_isStatic = false; ///< If true, the transform is static and doesn't move while entity is active.
         /// Behavior for this entity's transform when its parent's transform changes.
         AZ::OnParentChangedBehavior m_onParentChangedBehavior = AZ::OnParentChangedBehavior::Update;
+
+        // carbonated begin aoreshko game_specific_1
```
Comment (aefimov3) [inline Code/Framework/AzFramework/AzFramework/Components/TransformComponent.h]: Please remove aoreshko here and the other places, we drop names from engine patches.

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.h (hunk 4 above)
Comment (aefimov3) [reply Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.h]: We drop names from engine code, just short branch name.

File: Code/Framework/AzCore/AzCore/Component/TransformBus.h [hunk 11]
```diff
@@ -302,6 +302,13 @@ namespace AZ
 
         //! Set the behavior at runtime when this entity's parent's transform changes.
         virtual void SetOnParentChangedBehavior([[maybe_unused]] OnParentChangedBehavior onParentChangedBehavior) {}
+
+        // carbonated begin aoreshko game_specific_1
```
Comment (aefimov3) [inline Code/Framework/AzCore/AzCore/Component/TransformBus.h]: Is "game_specific_1" the branch name?
It is supposed to be an unique git branch name like feature/game_specific1. Game specific can be a comment (justification text), but the branch name should be more meaningful. Naming them game_specific_1, game_specific_2 makes not so much sense.

File: Code/Framework/AzFramework/AzFramework/Components/TransformComponent.cpp [hunk 12]
```diff
@@ -707,6 +707,16 @@ namespace AzFramework
         provided.push_back(AZ_CRC("TransformService", 0x8ee22c50));
     }
 
+    // carbonated begin enable_catbonated_1: Methids called from o2de-gruber
```
Comment (vmednonogov) [inline Code/Framework/AzFramework/AzFramework/Components/TransformComponent.cpp]: Methids -> Methods
o2de-gruber -> here and in other places

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.cpp [hunk 13]
```diff
@@ -366,6 +366,49 @@ namespace PhysX
...
 
+    // carbonated begin enable_catbonated_1: Methids called from o2de-gruber
+ #if defined(CARBONATED)
+    // Pilfered/inspired from SystemComponent::UpdateMaterialSelection
+    void CharacterControllerComponent::SetMaterialByName(uint32_t index, const AZStd::string& name)
+    {
+        AZ_Assert(m_characterConfig, "Character Config is null!");
+
+        Physics::MaterialAsset::MaterialProperties materialConfig;
+        //Physics::MaterialAsset::MaterialFromAssetConfiguration materialConfig;
```
Comment (vmednonogov) [inline Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.cpp]: Please remove this commented line or add TODO if you want to use it later

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.cpp [hunk 14]
```diff
@@ -366,6 +366,49 @@ namespace PhysX
...
+
+        const Physics::MaterialLibraryAsset* materialLibrary = m_characterConfig->m_materialSelection.GetMaterialLibraryAssetData();
+        //const Physics::MaterialLibraryAsset* materialLibrary = m_characterConfig->m_materialSelection.GetMaterialLibraryAssetData();
```
Comment (vmednonogov) [inline Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.cpp]: Please remove this commented line or add TODO if you want to use it later

**Pull request closed without merging**